softlayer/resource_softlayer_global_ip.go

```diff
@@ -124,7 +125,22 @@ func resourceSoftLayerGlobalIpUpdate(d *schema.ResourceData, meta interface{}) e
 		return fmt.Errorf("Not a valid global ip ID, must be an integer: %s", err)
 	}
 
-	_, err = service.Id(globalIpId).Route(sl.String(d.Get("routes_to").(string)))
+	routes_to := d.Get("routes_to").(string)
+	if strings.Contains(routes_to, ":") && len(routes_to) != 39 {
+		if colonCount := strings.Count(routes_to, ":"); colonCount != 7 {
+			routes_to = strings.Replace(routes_to, "::", "::"+strings.Repeat(":", 7-colonCount), 1)
+		}
+
+		parts := strings.Split(routes_to, ":")
+		for x, s := range parts {
+			parts[x] = fmt.Sprintf("%04s", s)
+		}
+
+		routes_to = strings.Join(parts, ":")
+		d.Set("routes_to", routes_to)
+	}
+
+	_, err = service.Id(globalIpId).Route(sl.String(routes_to))
```

Comment:
A couple of things:

1. We need to do this conversion for Create as well, no?
1. For Read, we should use net.ParseIP output and compare with net.ParseIP output of current value in `d`. Only if they are different, do we set `d` with new value (net.ParseIP output).

Good job in improving the expansion algorithm from the #golang channel. I would prefer though to do it this slightly faster way (replaces 3 calls to strings funcs with 1):

```go
        parts := strings.Split(routes_to, ":")
	for x, s := range parts {
		if s == "" {
                        zeroes := 9 - len(parts)
			parts[x] = strings.Repeat("0000:", zeroes)[:(zeroes*4)+(zeroes-1)]
		} else {
			parts[x] = fmt.Sprintf("%04s", s)
		}
	}
```